src/ndpp_ops.F90

```diff
@@ -0,0 +1,1083 @@
+    real(8)       :: thin_tol      ! Thinning tolerance used in lib, discarded
+    integer       :: NEin, iE      ! Number of incoming energies and the index
+    real(8), allocatable :: temp_outgoing(:,:) ! Temporary storage of scatt data
+    integer       :: NP            ! Number of precursors groups
+
+    ! determine path, record length, and location of table
+    filename      = listing % path
+    filetype      = listing % filetype
+    location      = listing % location
+
+    ! Check if ACE library exists and is readable
+    inquire(FILE=filename, EXIST=file_exists, READ=readable)
+    if (.not. file_exists) then
+      call fatal_error("NDPP library '" // trim(filename) // &
+           "' does not exist!")
+    elseif (readable(1:3) == 'NO') then
+      call fatal_error("NDPP library '" // trim(filename) // &
+           "' is not readable! Change file permissions with chmod command.")
+    end if
+
+    if (listing % filetype == ASCII) then
+      ! =======================================================================
+      ! READ NDPP DATA IN ASCII FORMAT
+
+      ! Find location of table
+      open(UNIT=in, FILE=filename, STATUS='old', ACTION='read')
+      rewind(UNIT=in)
+      do i = 1, location - 1
+        read(UNIT=in, FMT=*)
+      end do
+
+      ! Read the header information to make sure this is the correct file
+      read(UNIT=in, FMT='(A20,1PE20.12,I20,A20)') name, kT, NG
+      dkT = abs(kT - listing % kT) / kT
+      if ((adjustl(trim(name)) /= listing % name) .or. (dkT > 0.001_8)) then
+        call fatal_error("NDPP library '" // trim(filename) // &
+             "' does not contain the correct data set where expected!")
+      end if
+      ! Get the energy bins (not checking, will assume from here on out we have
+      ! the right data set)
+      allocate(energy_bins(NG + 1))
+      read(UNIT=in, FMT=*) energy_bins
+      ! The next line is scatt_type, lib_order, nuscatter, chi_present
+      read(UNIT=in, FMT='(I20,I20,I20,I20)') scatt_type, lib_order, &
+        nuscatter, chi_present
+      ! Finally, mu_bins, thin_tol
+      read(UNIT=in, FMT='(I20,1PE20.12)') mu_bins, thin_tol
+
+      ! set lib_order to the right number for allocating the outgoing array
+      if (scatt_type == SCATT_TYPE_LEGENDRE) lib_order = lib_order + 1
+
+      ! Start with elastic data
+      ! Get Ein information
+
+      ! First find our data length
+      read(UNIT=in, FMT='(I20)') NEin
+      ! Now allocate all that will be filled
+      allocate(this % el_Ein(NEin))
+      allocate(this % el_Ein_srch(NG + 1))
+      allocate(this % el(NEin))
+
+      ! Now read in el_Ein and el_Ein_srch
+      read(UNIT=in, FMT=*) this % el_Ein
+      read(UNIT=in, FMT=*) this % el_Ein_srch
+
+      ! Get the elastic moments themselves
+      do iE = 1, NEin
+        ! get gmin and gmax
+        read(UNIT=in, FMT='(I20,I20)') gmin, gmax
+
+        if ((gmin > 0) .and. (gmax > 0)) then
```

Comment:
This conditional fails when the NDPP data has no distribution for an incoming energy point.  This can theoretically occur (though it is not expected) when, for example, the only inelastic reaction channel present at an energy is described by a law that NDPP does not cover yet (because its crazy rare and I havent seen it yet), or sometimes if the ACE data is not self consistent.

During the tallying process (specifically within generate_ndpp_distrib_*n) this is checked.  It needs to be checked for other purposes besides this rare case, such as if the next data point is not allocated (because you are exactly at the upper bound) so there sadly isnt much time savings to be had by just checking up front.